l3_terrain_model_generator/src/plugins/std/generator/occupancy_map_generator.cpp

```diff
@@ -84,26 +89,112 @@ void OccupancyMapGenerator::update(const Timer& timer, UpdatedHandles& updates,
     return;
   }
 
-  // convert grid map to occupancy grid
-  grid_map::GridMapRosConverter::toOccupancyGrid(grid_map, layer_, ref_height + min_height_,
-                                                 ref_height + max_height_, occupancy_map);
+  // transform to reference frame if required
+  if (use_z_ref_frame_ && transform_to_ref_frame_)
+  {
+    // create a new grid map with the same size and resolution
+    grid_map::GridMap transformed_grid_map;
+    transformed_grid_map.setGeometry(grid_map.getLength(), grid_map.getResolution(), grid_map.getPosition());
+    transformed_grid_map.setStartIndex(grid_map.getStartIndex());
+    transformed_grid_map.setFrameId(grid_map.getFrameId());
+    transformed_grid_map.setTimestamp(grid_map.getTimestamp());
+    transformed_grid_map.add(layer_, grid_map.get(layer_));
+
+    // create debug map if required
+    if (publish_debug_map_)
+    {
+      debug_grid_map_ = grid_map::GridMap();
+      debug_grid_map_.setGeometry(grid_map.getLength(), grid_map.getResolution(), grid_map.getPosition());
+      debug_grid_map_.setStartIndex(grid_map.getStartIndex());
+      debug_grid_map_.setFrameId(grid_map.getFrameId());
+      debug_grid_map_.setTimestamp(grid_map.getTimestamp());
+      debug_grid_map_.add("debug_threshold_plane");
+    }
+
+    // Pre-compute plane coefficients
+    Eigen::Vector3d t = ref_pose.translation();
+    Eigen::Vector3d n = ref_pose.rotation().col(2);
+
+    // Plane equation: z = a*x + b*y + c
+    double a = -n.x() / n.z();
+    double b = -n.y() / n.z();
+    double c = t.z() + (n.x() / n.z()) * t.x() + (n.y() / n.z()) * t.y();
 
-  occupancy_map.info.origin.position.z = ref_height;
+    // Collect valid cells
+    std::vector<grid_map::Index> indices;
+    std::vector<grid_map::Position> positions;
+    indices.reserve(transformed_grid_map.getSize().prod());
+    positions.reserve(transformed_grid_map.getSize().prod());
+
+    for (grid_map::GridMapIterator it(transformed_grid_map); !it.isPastEnd(); ++it)
+    {
+      if (!transformed_grid_map.isValid(*it, layer_))
+        continue;
+
+      grid_map::Position pos;
+      transformed_grid_map.getPosition(*it, pos);
+
+      indices.push_back(*it);
+      positions.push_back(pos);
+    }
+
+    // Vectorized Z computation; We avoid heap allocations (using big Eigen Matrix) for small N
+    const size_t N = positions.size();
+    Eigen::VectorXd Z(N);
+
+    for (size_t i = 0; i < N; ++i)
+    {
+      Z(i) = a * positions[i].x() + b * positions[i].y() + c;
+    }
+
+    // Write back to grid map
+    for (size_t i = 0; i < N; ++i)
+    {
+      transformed_grid_map.at(layer_, indices[i]) -= Z(i);
```

Comment:
See previous comment: Avoid calls that do string lookups in loops iterating over cells.